src/ros_tcp_endpoint/tcp_sender.py

```diff
@@ -19,108 +19,103 @@
         # if we have a valid IP at this point, it was overridden locally so always use that
-        self.ip_is_overridden = (self.unity_ip != '')
         self.timeout_on_connect = timeout_on_connect
         self.timeout_on_send = timeout_on_send
         self.timeout_on_idle = timeout_on_idle
         self.queue = Queue()
-        sender_thread = threading.Thread(target=self.sender_loop)
-        # Exit the server thread when the main thread terminates
-        sender_thread.daemon = True
-        sender_thread.start()
-
-    def handshake(self, incoming_ip, data):
-        message = UnityHandshake._request_class().deserialize(data)
-        self.unity_port = message.port
-        if not self.ip_is_overridden:
-            # hello Unity, we'll talk to you from now on
-            if message.ip == '':
-                # if the message doesn't specify an IP, just talk back to the incoming IP address
-                self.unity_ip = incoming_ip
-            else:
-                # otherwise Unity has set an IP override, so use that
-                self.unity_ip = message.ip
-        print("ROS-Unity Handshake received, will connect to {}:{}".format(self.unity_ip, self.unity_port))
-        return UnityHandshakeResponse(self.unity_ip)
+        self.sender_id = 1
+        
+        # variables needed for matching up unity service requests with responses
+        self.next_srv_id = 1001
+        self.srv_lock = threading.Lock()
+        self.services_waiting = {}
 
     def send_unity_error(self, error):
         self.send_unity_message("__error", RosUnityError(error))
 
     def send_unity_message(self, topic, message):
-        if self.unity_ip == '':
-            print("Can't send a message, no defined unity IP!".format(topic, message))
-            return
-
         serialized_message = ClientThread.serialize_message(topic, message)
         self.queue.put(serialized_message)
 
     def send_unity_service(self, topic, service_class, request):
-        if self.unity_ip == '':
-            print("Can't send a message, no defined unity IP!".format(topic, request))
-            return
-
-        serialized_message = ClientThread.serialize_message(topic, request)
+        request_bytes = BytesIO()
+        request.serialize(request_bytes)
+        result = [None]
+        condition = threading.Condition()
         
-        try:
-            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            s.settimeout(self.timeout_on_connect)
-            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-            s.connect((self.unity_ip, self.unity_port))
-            s.settimeout(self.timeout_on_send)
-            s.sendall(serialized_message)
-
-            destination, data = ClientThread.read_message(s)
+        with self.srv_lock:
+            srv_id = self.next_srv_id
+            self.next_srv_id+=1
+            self.services_waiting[srv_id] = (condition, result)
```

Comment:
This `(condition, result)` tuple looks like it's structured enough to be upgraded to a `NamedTuple` or a class

Reply:
Created a new class ThreadPauser for this.